ksqldb-engine/src/main/java/io/confluent/ksql/function/UdafFactoryInvoker.java

```diff
@@ -63,13 +62,12 @@
     if (!Modifier.isStatic(method.getModifiers())) {
       throw new KsqlException("UDAF factory methods must be static " + method);
     }
-    final UdafTypes types = new UdafTypes(method, functionName.text(), typeParser);
+    final UdafTypes types = new UdafTypes(method, functionName, typeParser);
     this.functionName = Objects.requireNonNull(functionName);
     this.aggregateArgType = Objects.requireNonNull(types.getAggregateSchema(aggregateSchema));
     this.aggregateReturnType = Objects.requireNonNull(types.getOutputSchema(outputSchema));
     this.metrics = Objects.requireNonNull(metrics);
-    this.params =
-        Collections.singletonList(types.getInputSchema(Objects.requireNonNull(inputSchema)));
+    this.params = types.getInputSchema(Objects.requireNonNull(inputSchema));
```

Comment:
this doesn't sit right with me. The initialization literals are not part of the parameters, they are something separate that initialize the aggregator itself. I'm wondering how this will bite us in the future - one place that comes to mind is when we describe the function I think it'll now include the initialization parameters when it shouldn't.

Another problem might come from handling of the schema provider annotations, which currently take in a list of the parameters (not including the initialization literals). I don't know whether they should or shouldn't, but it seems wrong to lump them in together.

Reply:
> The initialization literals are not part of the parameters

Actually, it is part of the parameters from a users point of view.  Think about how this is being invoked:

```sql
LATEST_BY_OFFSET(Col0, 2)
```

The above takes _two_ parameters, not _one_.  Yes, the first is the parameter that will be passed to the aggregator on each agg call, and any after that must be literals and are passed to the aggregators init call.  But they are _all_ parameters!

> I'm wondering how this will bite us in the future

We need to overhaul out UDAF framework at some point to support multiple input _expressions_, not just _constants_. This is needed to implement a proper topN UDAF for example. However, when we come to do this we will need to change all this code anyway. So I don't think this is painting us into a corner.

> one place that comes to mind is when we describe the function I think it'll now include the initialization parameters when it shouldn't.

I disagree, I think it _should_ include the init params. The fact that describe currently excludes the init params is probably very confusing!  Take the second example `LATEST_BY_OFFSET` method in the description above. The one with two parameters.  Describing it only lists one parameter, which is wrong. 

> Another problem might come from handling of the schema provider annotations, which currently take in a list of the parameters (not including the initialization literals)

This is accounted for in the code.  The init params are added to end of the param described by the schema annotation.

My conclusion after looking into making this change is that it works for now and allows users to create variants of UDAFs that take constants as additional parameters, like the enhancement to `LATEST_BY_OFFSET` @sarwarbhuiyan wants to raise in a PR.  Longer term, we need a better solution. We desperately need UDAF support for schema provider functions, not just annotations. But this is a much bigger piece of work, and requires more thought, a KLIP, and resourcing.  This improves things in the short term, without hurting the long term, IMHO.